File: problems/002-two-sum/go/optimized.go
package main

// Same single-pass hash map as clean, with one change: the map is sized to the
// input up front. A Go map grows by rehashing every key into a new, larger
// bucket array each time it outgrows its load factor, so a map built by
// repeated insertion re-hashes an n-element input roughly log2(n/8) times.
// Handing make() the final size pays for the buckets once. Same O(n) time and
// O(n) space as clean -- this is a constant-factor win, not a class change.
func optimized(c map[string]any) any {
	nums := c["nums"].([]any)
	target := c["target"].(float64)
	seen := make(map[float64]int, len(nums))
	for i, v := range nums {
		n := v.(float64)
		if j, ok := seen[target-n]; ok {
			return []int{j, i}
		}
		seen[n] = i
	}
	return []int{}
}

File: problems/002-two-sum/dart/optimized.dart
// O(n) time is unbeatable for two-sum, so "optimized" is a constant-factor win
// over clean, not a complexity one. clean does two hash probes per element --
// containsKey, then [] to read the index. This does ONE: a single lookup whose
// null return means "absent", distinguished from a real index-0 hit by the null
// itself rather than a second probe. Readability takes the hit that null-vs-zero
// has to be reasoned about. O(n) time, O(n) space.
dynamic solve(Map<String, dynamic> c) {
  final nums = (c['nums'] as List).cast<num>();
  final target = c['target'] as num;
  final seen = <num, int>{};
  for (var i = 0; i < nums.length; i++) {
    final j = seen[target - nums[i]];
    if (j != null) return [j, i];
    seen[nums[i]] = i;
  }
  return [];
}
